components/dashboard/src/projects/ProjectVariables.tsx

```diff
@@ -0,0 +1,132 @@
+                <ItemsList>
+                    <Item header={true} className="grid grid-cols-4 items-center">
+                        <ItemField>Name</ItemField>
+                        <ItemField>Value</ItemField>
+                        <ItemField>Visible in Workspaces?</ItemField>
+                        <ItemField></ItemField>
+                    </Item>
+                    {envVars.map(variable => {
+                        return <Item className="grid grid-cols-4 items-center">
+                            <ItemField>{variable.name}</ItemField>
+                            <ItemField>****</ItemField>
+                            <ItemField>{variable.censored ? 'Hidden' : 'Visible'}</ItemField>
+                            <ItemField className="flex justify-end">
+                                <ItemFieldContextMenu menuEntries={[
+                                    {
+                                        title: 'Delete',
+                                        customFontStyle: 'text-red-600 dark:text-red-400 hover:text-red-800 dark:hover:text-red-300',
+                                        onClick: () => deleteEnvVar(variable.id),
+                                    },
+                                ]} />
+                            </ItemField>
+                        </Item>
+                    })}
+                </ItemsList>
+            </>
+        }
+    </ProjectSettingsPage>;
+}
+
+function AddVariableModal(props: { project?: Project, onClose: () => void }) {
+    const [ name, setName ] = useState<string>("");
+    const [ value, setValue ] = useState<string>("");
+    const [ censored, setCensored ] = useState<boolean>(true);
+    const [ error, setError ] = useState<Error | undefined>();
+
+    const addVariable = async () => {
+        if (!props.project) {
+            return;
+        }
+        try {
+            await getGitpodService().server.setProjectEnvironmentVariable(props.project.id, name, value, censored);
+            props.onClose();
+        } catch (err) {
+            setError(err);
+        }
+    }
+
+    return <Modal visible={true} onClose={props.onClose} onEnter={() => { addVariable(); return false; }}>
+        <h3 className="mb-4">Add Variable</h3>
+        <div className="border-t border-b border-gray-200 dark:border-gray-800 -mx-6 px-6 py-4 flex flex-col">
+            <AlertBox><strong>Project variables might be accessible by anyone with read access to your repository.</strong><br/>Secret values can be exposed if they are printed in the logs, persisted to the file system, or made visible in workspaces.</AlertBox>
+            {error && <div className="bg-gitpod-kumquat-light rounded-md p-3 text-gitpod-red text-sm mb-2">
+                {error}
+            </div>}
+            <div className="mt-8">
+                <h4>Name</h4>
+                <input autoFocus className="w-full" type="text" name="name" value={name} onChange={e => setName(e.target.value)} />
```

Comment:
Yes please - I was able to add a variable with no name or value :)

Reply:
Good catch! We should indeed validate variable names here (for values though, pretty much any text is valid, including awkward symbols or leaving an empty string -- safely escaping/unescaping values is our responsibility).

I'll add that validation to the DB code, because client-side validation doesn't make sense here (because it still allows API users, or any module in our code, to use illegal values).

Reply:
Update: Seems like the specifications accept "any character except `=` or `NUL`" as valid environment variable names (including non-printable characters), but for compatibility reasons, it seems safer indeed to allow only:
- alphabeticals (upper/lowercase)
- digits (except as first character)
- underscores

Source: https://stackoverflow.com/a/2821183/3461173

(TL;DR The user env var name validation is still correct 😇 although the `+` is unnecessary)

Reply:
Validation implemented ✅